### PR TITLE
doc: update lifecycle.md with transition policies

### PR DESCRIPTION
**Pull request by jonathan-gramain**

### CODE REVIEW

File: docs/lifecycle.md [hunk 1]
```diff
@@ -1,24 +1,38 @@
...
+On versioned buckets, it can apply to current version data, as well as
+noncurrent versions, depending on the rule.
+
+AWS-specific transitions, e.g. to STANDARD_IA or GLACIER storage
+classes is not supported through zenko, if needed, it has to be set
+directly on AWS.
+
+## Design
 
 The implementation of lifecycle in S3 connector comprises several
```
Comment (rahulreddy) [inline docs/lifecycle.md]: It's better to remove S3 Connector/Zenko references and keep it open for both

File: docs/lifecycle.md [hunk 2]
```diff
@@ -120,32 +156,80 @@ buckets are processed.
...
 The resulting zookeeper list should contain a current view of all
 buckets that have lifecycle enabled, up to the point where the
 metadata log has been processed.
 
-### Conductor
+Note that in the future, we may request the MongoDB database to get
+the list of buckets with lifecycle configuration instead of using
+zookeeper nodes.
+
+### Replication Queue Processor
```
Comment (rahulreddy) [inline docs/lifecycle.md]: `### Replication Queue Processor (RQP)`

File: docs/lifecycle.md [hunk 3]
```diff
@@ -120,32 +156,80 @@ buckets are processed.
...
+
+The RQP will process replication entries the same way than done for
+regular CRR, although they may come from transition policies
+processing (generated by the Lifecycle Producer).
+
+### Replication Status Processor
```
Comment (rahulreddy) [inline docs/lifecycle.md]: `### Replication Status Processor (RSP)`

**Review by rahulreddy — CHANGES_REQUESTED** [review]
I think we need to change the names of `Replication Status Processor` and `Replication Queue Processor` to something more generic.

**Comment by rahulreddy** [comment]
@bert-e wait

**Comment by jonathan-gramain** [comment]
@rahulreddy I agree on changing the names, maybe something relating to a "data mover" or "data workflow"?

We also would like to rename the lifecycle components, as "producer" and "consumer" here is not very meaningful, we had though about "processor", e.g. "bucket processor" and "object processor", which also maps well to the topic names "bucket-tasks" and "object-tasks" which they respectively consume.

**Comment by rahulreddy** [comment]
I didn't mean to change right now, right now. It's a thought for later.

**Comment by jonathan-gramain** [comment]
I changed the naming of the lifecycle components in the doc. If people are good with the naming I think we can do this change before RC3 so we can move on with a better naming scheme, otherwise I fear it's never going to happen...

File: docs/lifecycle.md [hunk 4]
```diff
@@ -1,26 +1,40 @@
...
+Transition policies allow to transition object data from one location
+to another one automatically, usually to a slower but cheaper location
+for old or infrequently-accessed data.
 
-Most components rely on Kafka and Zookeeper to pass state to each
+On versioned buckets, it can apply to current version data, as well as
+noncurrent versions, depending on the rule.
+
+AWS-specific transitions, e.g. to STANDARD_IA or GLACIER storage
+classes is not supported through zenko, if needed, it has to be set
```
Comment (dora-korpar) [inline docs/lifecycle.md]: zenko -> Zenko

File: docs/lifecycle.md [hunk 5]
```diff
@@ -108,6 +128,23 @@ See AWS specs for details on the protocol format.
 Those calls essentially manage bucket attributes related to lifecycle
 behavior, which are stored as part of bucket metadata.
 
+#### Transition Policies support in CloudServer
```
Comment (dora-korpar) [inline docs/lifecycle.md]: looks like you're going for non-proper nouns not being capitalized in the #### sections, so
Policies -> policy

File: docs/lifecycle.md [hunk 6]
```diff
@@ -108,6 +128,23 @@ See AWS specs for details on the protocol format.
...
+
+We introduce an optional per-object *preferred read location*. If set,
+it is used to determine which location to read the object from on a
+GET request. It overrides the *preferred read location* defined in the
+bucket replication configuration (if any).
+
+The attribute may be present in the `replicationInfo` section of
+object metadata, as a `"preferred": "<location>"` property, where
+`<location>` is one of the replication targets listed in
+`replicationInfo` that should be in `COMPLETED` state in order to make
```
Comment (dora-korpar) [inline docs/lifecycle.md]: I'd put a period after `replicationInfo` and start next sentence with 'It should be in the 'COMPLETED' state...'

File: docs/lifecycle.md [hunk 7]
```diff
@@ -120,32 +157,80 @@ buckets are processed.
...
 the metadata log:
 
 * when processing a log entry that is an update of bucket attributes,
   look at whether any lifecycle action is enabled for that bucket:
-  * if any action is enabled, create the zookeeper node for this
-    owner/bucket (ignore if already exists)
-  * if no action is enabled, delete the zookeeper node for this
-    owner/bucket (ignore if not found)
+
+  * if any action is enabled, create the zookeeper node(ignore if
```
Comment (dora-korpar) [inline docs/lifecycle.md]: node(ignore -> node (ignore

File: docs/lifecycle.md [hunk 8]
```diff
@@ -120,32 +157,80 @@ buckets are processed.
...
+  the zookeeper node (ignore if not found).
 
 The resulting zookeeper list should contain a current view of all
 buckets that have lifecycle enabled, up to the point where the
 metadata log has been processed.
 
-### Conductor
+Note that in the future, we may request the MongoDB database to get
+the list of buckets with lifecycle configuration instead of using
+zookeeper nodes.
```
Comment (dora-korpar) [inline docs/lifecycle.md]: Zookeeper

File: docs/lifecycle.md [hunk 9]
```diff
@@ -120,32 +157,80 @@ buckets are processed.
...
+
+### Replication Queue Processor (RQP)
+
+The RQP will process replication entries the same way than done for
```
Comment (dora-korpar) [inline docs/lifecycle.md]: "... entries in the same way as regular CRR..."

File: docs/lifecycle.md [hunk 10]
```diff
@@ -175,66 +260,82 @@ only the specific range belonging to this prefix. For this we can add
...
 this range.
 
-### Producer
+### Lifecycle Bucket Processor (LBP)
 
-The producer's responsibility is to list objects in lifecycled
-buckets, and publish to kafka queues the actual lifecycle actions to
-execute on objects. For now, supported actions are expiration actions.
+LBP's responsibility is to list objects in lifecycled buckets, and
+publish to kafka queues the actual lifecycle actions to execute on
```
Comment (dora-korpar) [inline docs/lifecycle.md]: Kafka

File: docs/lifecycle.md [hunk 11]
```diff
@@ -175,66 +260,82 @@ only the specific range belonging to this prefix. For this we can add
...
+objects. For now, supported actions are expiration actions.
```
Comment (dora-korpar) [inline docs/lifecycle.md]: Not true anymore! :)

File: docs/lifecycle.md [hunk 12]
```diff
@@ -175,66 +260,82 @@ only the specific range belonging to this prefix. For this we can add
...
 
-More specifically, the producer is part of a kafka consumer group that
+More specifically, the LBP is part of a kafka consumer group that
```
Comment (dora-korpar) [inline docs/lifecycle.md]: Kafka

File: docs/lifecycle.md [hunk 13]
```diff
@@ -175,66 +260,82 @@ only the specific range belonging to this prefix. For this we can add
...
-    listing and processing of this bucket.
+    finished listing. Then another LBP instance will keep going with
+    the listing and processing of this bucket.
+
   * For each object listed, match it against the lifecycle rules
-    logic. For each lifecycle rule matching (e.g. expiry), publish to
-    the *backbeat-lifecycle-object-tasks* kafka topic an entry that
-    contains the required info for the consumer to execute the
-    lifecycle action (see [object tasks](#object-tasks) message format).
+    logic. For each lifecycle rule matching, publish to the relevant
```
Comment (dora-korpar) [inline docs/lifecycle.md]: "...publish the action(s) to take to the relevant topic(s):"

File: docs/lifecycle.md [hunk 14]
```diff
@@ -175,66 +260,82 @@ only the specific range belonging to this prefix. For this we can add
...
+#### Note 2
 
-Periodically (every minute in the current default settings) the
-internal consumer of the bucket tasks topic publishes its current
-committed offset and the latest topic offset in zookeeper under
-`/[chroot_path]/lifecycle/run/backlog-metrics/...`, for checking in
-the conductor.
+We may do search requests on MongoDB to query the exact list of
+objects where a lifecycle rule applies, instead of filtering based on
+the full listing result, for a better efficiency.
```
Comment (dora-korpar) [inline docs/lifecycle.md]: no "a"

File: docs/lifecycle.md [hunk 15]
```diff
@@ -108,6 +128,23 @@ See AWS specs for details on the protocol format.
...
 
+#### Transition policies support in CloudServer
+
+Transition policies rely on the *preferred read location* for an
+object.
+
+We introduce an optional per-object *preferred read location*. If set,
+it is used to determine which location to read the object from on a
+GET request. It overrides the *preferred read location* defined in the
+bucket replication configuration (if any).
```
Comment (bennettbuchanan) [inline docs/lifecycle.md]: Is is possible to set this in the bucket replication config?

File: docs/lifecycle.md [hunk 16]
```diff
@@ -117,35 +154,87 @@ It is extended for lifecycle management, to maintain a list of buckets
...
-  the zookeeper node for that owner/bucket (ignore if not found).
+  the Zookeeper node (ignore if not found).
+
+The resulting Zookeeper list should contain a current view of all
+buckets that have an attached and enabled lifecycle configuration, up
+to the point where the metadata log has been processed.
+
+Note that in the future, we may request the MongoDB database to get
+the list of buckets with lifecycle configuration instead of using
+Zookeeper nodes.
```
Comment (bennettbuchanan) [inline docs/lifecycle.md:194]: What would prevent us from implementing this MongoDB usage off-the-bat?

**Review by bennettbuchanan — APPROVED** [review]

File: docs/lifecycle.md (hunk 15 above)
Comment (jonathan-gramain) [reply docs/lifecycle.md]: Yes, it's done by appending `:preferred_read` to one of the target locations in the `StorageClass` field.

File: docs/lifecycle.md (hunk 16 above)
Comment (jonathan-gramain) [reply docs/lifecycle.md:194]: We need the proper clients to be configured to query mongodb directly and extract the full list of buckets from there.

I recall the zookeeper list method has been implemented to only consider buckets that have lifecycle attached to them for efficiency, but I am thinking this is not that useful and brings more complexity that could be removed eventually.

**Review by rahulreddy — APPROVED** [review]

**Comment by jonathan-gramain** [comment]
@bert-e approve

**Pull request merged**